src/main/java/net/fabricmc/loom/providers/MappingsProvider.java

```diff
@@ -24,39 +24,85 @@
+import java.net.URL;
 import java.nio.file.FileSystem;
 import java.nio.file.FileSystems;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
 import java.util.function.Consumer;
 
+import javax.annotation.Nullable;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.tools.ant.util.StringUtils;
 import org.gradle.api.Project;
+import org.gradle.api.artifacts.Dependency;
 
 import net.fabricmc.loom.LoomGradleExtension;
 import net.fabricmc.loom.util.Constants;
 import net.fabricmc.loom.util.DependencyProvider;
 import net.fabricmc.loom.util.Version;
-import net.fabricmc.mappings.Mappings;
+import net.fabricmc.mapping.tree.TinyMappingFactory;
+import net.fabricmc.mapping.tree.TinyTree;
+import net.fabricmc.stitch.Command;
 import net.fabricmc.stitch.commands.CommandProposeFieldNames;
+import net.fabricmc.stitch.commands.tinyv2.CommandMergeTinyV2;
+import net.fabricmc.stitch.commands.tinyv2.CommandReorderTinyV2;
 
-//TODO fix local mappings
-//TODO possibly use maven for mappings, can fix above at the same time
 public class MappingsProvider extends DependencyProvider {
 	public MinecraftMappedProvider mappedProvider;
 
 	public String mappingsName;
 	public String minecraftVersion;
 	public String mappingsVersion;
 
-	public File MAPPINGS_DIR;
-	public File MAPPINGS_TINY_BASE;
-	public File MAPPINGS_TINY;
-	public File MAPPINGS_MIXIN_EXPORT;
+	private boolean isV2;
+	private Path mappingsDir;
+	private Path mappingsStepsDir;
+	private Path tinyMappingsWithoutEnums;
+	public File tinyMappings;
+	public File mappingsMixinExport;
+
+	public void clean() throws IOException {
+		tinyMappings.delete();
+		FileUtils.deleteDirectory(mappingsStepsDir.toFile());
+	}
+
+	public TinyTree getMappings() throws IOException {
+		return MappingsCache.INSTANCE.get(tinyMappings.toPath());
+	}
+
+	@Nullable
+	public TinyTree getMappingsOfVersion(Project project, String version) throws IOException {
+		Path migrateMappingsDir = mappingsDir.resolve("migrate");
+		Path localMappingsOfVersion = migrateMappingsDir.resolve(version + ".tiny");
 
-	public Mappings getMappings() throws IOException {
-		return MappingsCache.INSTANCE.get(MAPPINGS_TINY.toPath());
+		if (Files.exists(localMappingsOfVersion)) {
+			try (BufferedReader reader = Files.newBufferedReader(localMappingsOfVersion)) {
+				return TinyMappingFactory.loadWithDetection(reader);
+			}
+		} else {
+			//TODO: download unmerged v2 mappings once they are on the maven (v2-yarn)
+			String versionRelativePath = version.replace("+", "%2B").replace(" ", "%20");
+			String artifactUrl = "https://maven.fabricmc.net/net/fabricmc/yarn/" + versionRelativePath + "/yarn-" + versionRelativePath + ".jar";
+			File mappingsJar = File.createTempFile("migrateMappingsJar", ".jar");
+			project.getLogger().lifecycle(":downloading new mappings from " + artifactUrl);
+			FileUtils.copyURLToFile(new URL(artifactUrl), mappingsJar);
```

Comment:
DownloadUtil.downloadIfChanged ?

Reply:
Is there much point when we need to extract the mappings when we first download them, and therefore this only gets called when the mappings file doesn't exist at all?

Reply:
How often do we use/download this? and if we extract, the non-extracted file is still there, so the downloadifchanged logic can still work

Reply:
we redownload only if the file doesn't exist at all, see line 85 `if (Files.exists(localMappingsOfVersion)) {`